src/net/ossl.cc

```diff
@@ -1657,9 +1581,181 @@ class session : public enable_shared_from_this<session>, public session_impl {
     session_type _type;
     bool _eof = false;
     bool _shutdown = false;
+
+    friend int bio_write_ex(BIO* b, const char * data, size_t dlen, size_t * written);
+    friend int bio_read_ex(BIO* b, char * data, size_t dlen, size_t *readbytes);
+    friend long bio_ctrl(BIO * b, int ctrl, long num, void * data);
 };
+
+
+tls::session* unwrap_bio_ptr(void * ptr) {
+    return static_cast<tls::session*>(ptr);
+}
+
+tls::session* unwrap_bio_ptr(BIO * b) {
+    return unwrap_bio_ptr(BIO_get_data(b));
+}
+
+/// The 'ioctl' for BIO
+long bio_ctrl(BIO * b, int ctrl, long num, void * data) {
+    long ret = 1;
+
+    if (BIO_get_init(b) <= 0 && ctrl != BIO_C_SET_POINTER) {
+        return 0;
+    }
+
+    auto session = unwrap_bio_ptr(b);
+
+    switch(ctrl) {
+    case BIO_C_SET_POINTER:
+        if (BIO_get_init(b) <= 0) {
+            BIO_set_data(b, data);
+            BIO_set_init(b, 1);
+        } else {
+            ret = 0;
+        }
+        break;
+    case BIO_CTRL_GET_CLOSE:
+        ret = BIO_get_shutdown(b);
+        break;
+    case BIO_CTRL_SET_CLOSE:
+        BIO_set_shutdown(b, static_cast<int>(num));
+        break;
+
+    case BIO_CTRL_DUP:
+    case BIO_CTRL_FLUSH:
+        ret = 1;
+        break;
+    case BIO_CTRL_EOF:
+        ret = BIO_test_flags(b, BIO_FLAGS_IN_EOF) != 0;
+        break;
+    case BIO_CTRL_PENDING:
+        ret = static_cast<long>(session->_input.size());
+        break;
+    case BIO_CTRL_WPENDING:
+        ret = session->_output_pending.available() ? 0 : 1;
+        break;
+    default:
+        ret = 0;
+    }
+    return ret;
+}
+
+/// This is called when the BIO is created
+///
+/// It is important for this to be set, even if it doesn't do anything
+/// because the BIO init flag won't get automatically set to '1'.
+int bio_create(BIO*) {
+    return 1;
+}
+
+/// Handles writes to the BIO
+///
+/// This function will attempt to call _out.put() and store the future in
+/// _output_pending.  If _output_pending has not yet resolved, return '0'
+/// and set the retry write flag.
+int bio_write_ex(BIO* b, const char * data, size_t dlen, size_t * written) {
+    auto session = unwrap_bio_ptr(b);
+    BIO_clear_retry_flags(b);
+
+    if (!session->_output_pending.available()) {
+        BIO_set_retry_write(b);
+        return 0;
+    }
+
+    try {
+        size_t n;
+
+        if (!session->_output_pending.failed()) {
+            scattered_message<char> msg;
+            msg.append(std::string_view(data, dlen));
+            n = msg.size();
+            session->_output_pending = session->_out.put(std::move(msg).release());
+        }
+
+        if (session->_output_pending.failed()) {
+            std::rethrow_exception(session->_output_pending.get_exception());
```

Comment:
Are we expecting the future from `put` to be resolved here, so that it can be detected as failed, or could this just be an else block?

Reply:
I'm also a little confused as to why we'd want to throw in this method. From what I remember returning a -1 or something from this method tells openssl that this operation failed. 

Is it maybe because theres an unrecoverable error? Maybe another option would be to let `get()` or `put()` throw the next time they are called?

Reply:
Here it's just a check to see that if any part of the synchronous call to `put` fails, we catch it immediately, otherwise it'll get caught on the next `wait_for_output` call.

Reply:
@graphcareful This is wrapped in a try-catch so the exception won't leak.